packages/embedder/embedder.ts

```diff
@@ -0,0 +1,250 @@
+  //   * I don't love all overrides, I wish there was a better way
+  //     see: https://github.com/kysely-org/kysely/issues/872
+
+  const batchToQueue = await selectMetaToQueue(configuredModels, orgIds, itemCountToQueue)
+
+  if (!batchToQueue.length) {
+    console.log(`embedder: no new items to queue`)
+    return
+  }
+
+  const ejqHash: {
+    [key: string]: {
+      refUpdatedAt: Date
+    }
+  } = {}
+  const makeKey = (item: {objectType: string | null; refId: string | null}) =>
+    `${item.objectType}:${item.refId}`
+
+  const ejqValues = batchToQueue.map((item) => {
+    ejqHash[makeKey(item)] = {
+      refUpdatedAt: item.refUpdatedAt
+    }
+    return {
+      objectType: item.objectType,
+      refId: item.refId as string,
+      model: item.model,
+      state: 'queued' as DBInsert['EmbeddingsJobQueue']['state']
+    }
+  })
+
+  const ejqRows = await insertNewJobs(ejqValues)
+
+  ejqRows.forEach((item) => {
+    const {refUpdatedAt} = ejqHash[makeKey(item)]!
+    const score = new Date(refUpdatedAt).getTime()
+    redisClient.zadd('embedder:queue', score, item.id)
+  })
+
+  console.log(`embedder: queued ${batchToQueue.length} items`)
+}
+
+const dequeueAndEmbedUntilEmpty = async (modelManager: ModelManager) => {
+  while (true) {
+    const maybeRedisQItem = await redisClient.zpopmax('embedder:queue', 1)
+    if (maybeRedisQItem.length < 2) return // Q is empty, all done!
+
+    const [id, _] = maybeRedisQItem
+    if (!id) {
+      console.warn(`De-queued undefined item from embedder:queue`)
+      continue
+    }
+    const jobQueueId = parseInt(id, 10)
+    const jobQueueItem = await selectJobQueueItemById(jobQueueId)
+    if (!jobQueueItem) {
+      console.warn(`Unable to fetch EmbeddingsJobQueue.id = ${id}`)
+      continue
+    }
+
+    const metadata = await selectMetadataByJobQueueId(jobQueueId)
+    if (!metadata) {
+      await updateJobState(jobQueueId, 'failed', {
+        stateMessage: `unable to fetch metadata by EmbeddingsJobQueue.id = ${id}`
+      })
+      continue
+    }
+
+    let fullText = metadata?.embedText
+    try {
+      if (!fullText) {
+        fullText = await createEmbeddingTextFrom(jobQueueItem, dataLoader)
+      }
+    } catch (e) {
+      await updateJobState(jobQueueId, 'failed', {
+        stateMessage: `unable to create embedding text: ${e}`
+      })
+      continue
+    }
+
+    const wordCount = countWords(fullText)
+    for (const embeddingModel of modelManager.getEmbeddingModelsIter()) {
+      let embedText = fullText
+      const {maxInputTokens} = embeddingModel.getModelParams()
+      // we're using word count as an appoximation of tokens
+      if (wordCount * WORD_COUNT_TO_TOKEN_RATIO > maxInputTokens) {
+        try {
+          const generator = modelManager.getFirstGenerator()
+          if (!generator) throw new Error(`Generator unavailable`)
+          const textToSummarize = fullText.slice(0, generator.getModelParams().maxInputTokens)
```

Comment:
This is a hack.

I am "in hate" with the state of tokenization. The fullText to embed is too long to be embedded directly, so we want to summarize it first. The GPT also has a maximum context window length. It will throw an error if you give it too much text. So, we need to make it shorter. Ideally the API would allow us to tokenize text in order to know how long it is. That API doesn't exist, so we just hack it down to a value we'll know fits before summarizing it.

I added a comment.